### PR TITLE
Version 0.9.0

### PR DESCRIPTION
**Pull request by ubaldot**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by zklaus** [comment]
In the `requirements/host` section, you currently have `python >= {{ python_min }}`. This is not correct. It should be `python {{ python_min }}`. That is so that it is guaranteed that this package will be build/installed with the lowest python version. For the `requirements/run` section, it **is** correct to have `python >= {{ python_min }}`, so that the package can be installed into environments with newer python versions.

To address the errors you must rerender as instructed in the checklist in the description of this PR.

**Comment by ubaldot** [comment]
@conda-forge-admin please rerender

**Comment by ubaldot** [comment]
> In the `requirements/host` section, you currently have `python >= {{ python_min }}`. This is not correct. It should be `python {{ python_min }}`. That is so that it is guaranteed that this package will be build/installed with the lowest python version. For the `requirements/run` section, it **is** correct to have `python >= {{ python_min }}`, so that the package can be installed into environments with newer python versions.
> 
> To address the errors you must rerender as instructed in the checklist in the description of this PR.

@zklaus Thanks! 
Oh ok. I read this in the instructions:

We need to re-render when there are changes in the following parts of the feedstock:

1. Platform configuration (skip sections).
2. yum_requirements.txt or conda-forge.yml.
3. Updates in the build matrix due to new versions of Python, NumPy, PERL, R, etc.
4. Updates in conda-forge pinning that affect the feedstock.
5. Build issues that a feedstock configuration update will fix (follow us on [Zulip](https://conda-forge.zulipchat.com/) to know about those).

this is why I didn't re-rendered. However, I asked our bot to render now :)

**Comment by ubaldot** [comment]
@zklaus Sorry for bothering you once more, but there is a problem in the build. In the log it says:

```
  from typing import Self
ImportError: cannot import name 'Self' from 'typing' (/home/conda/feedstock_root/build_artifacts/dymoval_1732874748106/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/python3.10/typing.py)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/conda/feedstock_root/build_artifacts/dymoval_1732874748106/test_tmp/run_test.py", line 2, in <module>
    import dymoval
  File "/home/conda/feedstock_root/build_artifacts/dymoval_1732874748106/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/python3.10/site-packages/dymoval/__init__.py", line 4, in <module>
    from .dataset import (
  File "/home/conda/feedstock_root/build_artifacts/dymoval_1732874748106/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/python3.10/site-packages/dymoval/dataset.py", line 15, in <module>
    from typing_extensions import Self  # noqa
ModuleNotFoundError: No module named 'typing_extensions'
```

which is strange because I have it included in [pyproject.toml ](https://github.com/ubaldot/dymoval/blob/main/pyproject.toml)

Also, the github [actions](https://github.com/ubaldot/dymoval/actions/runs/12074969548) go through.

In the source I have:

```
try:
    from typing import Self
except ImportError:
    from typing_extensions import Self  # noqa
```

But from the logs it seems that `typing_extensions` package is not installed. 
What is happening?

**Comment by zklaus** [comment]
### Regarding the rerendering

Maybe the formulation in the docs is not the clearest one, but when you look at the list there are a few things that really are outside of the feedstock, like the global pinning. Rerendering also does any upgrades in "the machine", i.e. a new version of conda-smithy etc.

tldr: Just rerender everytime. It may be necessary due to reasons you cannot know from the feedstock or the PR alone. If it's not necessary, it won't do any harm.
 

### Regarding the "Self" import

This is precisely the point of the whole `python_min` dance. Noarch packages work by being built **once** in a architecture independent manner. That means there is only a single package that will be installed on all platforms (x86, x86_64, arm64, ...; linux, macos, win, ...), which you can see [here](https://anaconda.org/conda-forge/dymoval/files). This works for pure python packages well because they only have byte-code compiled, which is universal. However, this package also is used across all python versions and that's where things can get more tricky. Python has excellent backward compatibility, i.e. any newer python version will happily run code that has been written for an older python version (with the notable exception of the python 2 to python 3 transition).

But forward compatibility is harder. Since the new versions of python may introduce new features (d'oh), older versions can not always run code that was written for newer versions. That's why also noarch packages should announce the oldest version that they are compatible with.

In your case, you are using a feature that was introduced in python 3.11 (see https://docs.python.org/3/library/typing.html#typing.Self). Before, when you had `python >= {{ python_min }}` in your host requirements, the CI installed the latest python version (3.12) for building the package and everything worked out, **but** things would break for someone that installs the package in an environment with python 3.9 (which is `python_min` [at time of writing](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/d679e04b6513646b4429d81baddcced9a2c198b4/recipe/conda_build_config.yaml#L798).
Using `python {{ python_min }}` in the host requirements guarantees that the package is built with the minimal supported python version which has now uncovered this problem.

### What to do

You have two options: Either you make the code compatible with python 3.9, making your package installable in a larger number of environments. If it's only a single instance of using `typing.Self`, that may be feasible.

Or you can declare python 3.11 to be the actually supported minimal python for your package. You do that by including the following line as the very first line of your meta.yaml file.
```
{% set python_min = python_min | default("3.11") %}
```

Warning: That means you package will not be installable in environments together with versions of python <3.11.

**Comment by zklaus** [comment]
Ah sorry, I didn't read your comment carefully enough. I'll leave my previous comment here for future reference, but I think what you have to do is simply add `typing_extensions` to your requirements in the recipe. Remember that conda-forge recipes are handled separately from dependencies in pyproject.toml and similar places and must be updated "by hand".

**Comment by ubaldot** [comment]
> ### Regarding the rerendering
> Maybe the formulation in the docs is not the clearest one, but when you look at the list there are a few things that really are outside of the feedstock, like the global pinning. Rerendering also does any upgrades in "the machine", i.e. a new version of conda-smithy etc.
> 
> tldr: Just rerender everytime. It may be necessary due to reasons you cannot know from the feedstock or the PR alone. If it's not necessary, it won't do any harm.
> 

OK! You mean to run locally as described [here](https://conda-forge.org/docs/maintainer/updating_pkgs/#rerendering-with-conda-smithy-locally)?

> ### Regarding the "Self" import
> This is precisely the point of the whole `python_min` dance. Noarch packages work by being built **once** in a architecture independent manner. That means there is only a single package that will be installed on all platforms (x86, x86_64, arm64, ...; linux, macos, win, ...), which you can see [here](https://anaconda.org/conda-forge/dymoval/files). This works for pure python packages well because they only have byte-code compiled, which is universal. However, this package also is used across all python versions and that's where things can get more tricky. Python has excellent backward compatibility, i.e. any newer python version will happily run code that has been written for an older python version (with the notable exception of the python 2 to python 3 transition).
> 

Thanks for the pedagogic explanation. Appreciated! 

> But forward compatibility is harder. Since the new versions of python may introduce new features (d'oh), older versions can not always run code that was written for newer versions. That's why also noarch packages should announce the oldest version that they are compatible with.
> 
> In your case, you are using a feature that was introduced in python 3.11 (see https://docs.python.org/3/library/typing.html#typing.Self). Before, when you had `python >= {{ python_min }}` in your host requirements, the CI installed the latest python version (3.12) for building the package and everything worked out, **but** things would break for someone that installs the package in an environment with python 3.9 (which is `python_min` [at time of writing](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/d679e04b6513646b4429d81baddcced9a2c198b4/recipe/conda_build_config.yaml#L798). Using `python {{ python_min }}` in the host requirements guarantees that the package is built with the minimal supported python version which has now uncovered this problem.
> 


> ### What to do
> You have two options: Either you make the code compatible with python 3.9, making your package installable in a larger number of environments. If it's only a single instance of using `typing.Self`, that may be feasible.
> 
> Or you can declare python 3.11 to be the actually supported minimal python for your package. You do that by including the following line as the very first line of your meta.yaml file.
> 
> ```
> {% set python_min = python_min | default("3.11") %}
> ```
> 
> Warning: That means you package will not be installable in environments together with versions of python <3.11.

In reality I dropped support for 3.9 and I stated that the minimum version should be 3.10 in the pyproject.toml. Can I just use `{% set python_min = python_min | default("3.10") %` ?

**Comment by ubaldot** [comment]
> Ah sorry, I didn't read your comment carefully enough. I'll leave my previous comment here for future reference, but I think what you have to do is simply add `typing_extensions` to your requirements in the recipe. Remember that conda-forge recipes are handled separately from dependencies in pyproject.toml and similar places and must be updated "by hand".

Sorry, we overlapped in the answers, but I appreciated your pedagogic explanation. 
I added the extra dependency as you pointed out. Let's see what happens now.

**Comment by ubaldot** [comment]
@zklaus It went through all the checks, finally! 

Many thanks for the help! Let's see how much will it take before I find my package on anaconda.org :p

Just a some feedback, from this experience. The docs may be updated with the following:

1. It shall be always run the renderer. It would be nice also to explain what render is, or if it is done elsewhere, add a link to the feedstock maintainer page. 
2. Stress that the dependencies on pyproject.toml and meta.yaml shall match and that shall be done manually
3. In this [section](https://conda-forge.org/docs/maintainer/updating_pkgs/#example-workflow-for-updating-a-package) It is stated: 

> Navigate to https://github.com/conda-forge/ in your favorite web browser and click the fork button.

That should be `https://github.com/conda-forge/<feedstock>`.
